File: Fileallocation.c
  le
char fileNames[MAX][20];
int fileBlocks[MAX][MAX];
int fileSizes[MAX];
char typeMap[MAX][20];
int fileCount = 0;

// 🔹 Display Disk#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX 100

int totalBlocks;
char disk[MAX][20];   // stores file name or empty

// File Tab
void displayDisk() {
    printf("\nDisk: ");
    for (int i = 0; i < totalBlocks; i++) {
        if (strlen(disk[i]) == 0)
            printf("- ");
        else
            printf("%s ", disk[i]);
    }
    printf("\n");
}

// 🔹 Show Files
void showFiles() {
    printf("\nFile Table:\n");
    for (int i = 0; i < fileCount; i++) {
        printf("%s (%s) -> ", fileNames[i], typeMap[i]);
        for (int j = 0; j < fileSizes[i]; j++) {
            printf("%d ", fileBlocks[i][j]);
        }
        printf("\n");
    }
}

// 🔹 Contiguous Allocation
void contiguous(char name[], int size) {
    int start = -1, count = 0;

    for (int i = 0; i < totalBlocks; i++) {
        if (strlen(disk[i]) == 0) {
            if (count == 0) start = i;
            count++;

            if (count == size) {
                printf("Allocated: ");
                for (int j = start; j < start + size; j++) {
                    strcpy(disk[j], name);
                    fileBlocks[fileCount][j - start] = j;
                    printf("%d ", j);
                }

                strcpy(fileNames[fileCount], name);
                strcpy(typeMap[fileCount], "Contiguous");
                fileSizes[fileCount] = size;
                fileCount++;
                printf("\n");
                return;
            }
        } else {
            count = 0;
        }
    }

    printf("Contiguous allocation failed\n");
}

// 🔹 Linked Allocation
void linked(char name[], int size) {
    int count = 0;

    printf("Allocated: ");
    for (int i = 0; i < totalBlocks && count < size; i++) {
        if (strlen(disk[i]) == 0) {
            strcpy(disk[i], name);
            fileBlocks[fileCount][count++] = i;
            printf("%d ", i);
        }
    }

    if (count < size) {
        printf("\nLinked allocation failed\n");
        return;
    }

    strcpy(fileNames[fileCount], name);
    strcpy(typeMap[fileCount], "Linked");
    fileSizes[fileCount] = size;
    fileCount++;

    printf("\n");
}

// 🔹 Indexed Allocation
void indexed(char name[], int size) {
    int count = 0;
    int blocks[MAX];

    for (int i = 0; i < totalBlocks && count < size + 1; i++) {
        if (strlen(disk[i]) == 0) {
            strcpy(disk[i], name);
            blocks[count++] = i;
        }
    }

    if (count < size + 1) {
        printf("Indexed allocation failed\n");
        return;
    }

    printf("Index Block: %d\n", blocks[0]);
    printf("Data Blocks: ");

    for (int i = 1; i < count; i++) {
        printf("%d ", blocks[i]);
        fileBlocks[fileCount][i - 1] = blocks[i];
    }

    strcpy(fileNames[fileCount], name);
    strcpy(typeMap[fileCount], "Indexed");
    fileSizes[fileCount] = size;
    fileCount++;

    printf("\n");
}

// 🔹 Main Function
int main() {
    int choice, type, size;
    char name[20];

    printf("Enter total blocks: ");
    scanf("%d", &totalBlocks);

    // Initialize disk
    for (int i = 0; i < totalBlocks; i++) {
        disk[i][0] = '\0';
    }

    while (1) {
        printf("\n1. Create File\n");
        printf("2. Display Disk\n");
        printf("3. Show Files\n");
        printf("4. Exit\n");

        scanf("%d", &choice);

        if (choice == 4) break;

        if (choice == 2) {
            displayDisk();
            continue;
        }

        if (choice == 3) {
            showFiles();
            continue;
        }

        printf("Choose Allocation:\n");
        printf("1. Contiguous\n");
        printf("2. Linked\n");
        printf("3. Indexed\n");

        scanf("%d", &type);

        printf("Enter file name: ");
        scanf("%s", name);

        printf("Enter size: ");
        scanf("%d", &size);

        switch (type) {
            case 1: contiguous(name, size); break;
            case 2: linked(name, size); break;
            case 3: indexed(name, size); break;
            default: printf("Invalid choice\n");
        }
    }

    return 0;
}

File: fileallocation.java
import java.util.*;

class FileAllocation {

    static int totalBlocks = 20;
    static String[] disk = new String[totalBlocks];

    // 🔹 Contiguous Allocation
    static void allocateContiguous(String name, int size) {
        int start = -1, count = 0;

        for (int i = 0; i < totalBlocks; i++) {
            if (disk[i] == null) {
                if (count == 0) start = i;
                count++;

                if (count == size) {
                    for (int j = start; j < start + size; j++) {
                        disk[j] = name;
                    }
                    System.out.println(name + " allocated from block " + start + " to " + (start + size - 1));
                    return;
                }
            } else {
                count = 0;
            }
        }

        System.out.println("Contiguous allocation failed");
    }

    // 🔹 Linked Allocation
    static void allocateLinked(String name, int size) {
        List<Integer> blocks = new ArrayList<>();

        for (int i = 0; i < totalBlocks && blocks.size() < size; i++) {
            if (disk[i] == null) {
                disk[i] = name;
                blocks.add(i);
            }
        }

        if (blocks.size() < size) {
            System.out.println("Linked allocation failed");
            return;
        }

        System.out.println(name + " allocated at blocks: " + blocks);
    }

    // 🔹 Indexed Allocation
    static void allocateIndexed(String name, int size) {
        List<Integer> blocks = new ArrayList<>();

        for (int i = 0; i < totalBlocks && blocks.size() < size + 1; i++) {
            if (disk[i] == null) {
                disk[i] = name;
                blocks.add(i);
            }
        }

        if (blocks.size() < size + 1) {
            System.out.println("Indexed allocation failed");
            return;
        }

        int indexBlock = blocks.get(0);
        List<Integer> fileBlocks = blocks.subList(1, blocks.size());

        System.out.println(name + " index block: " + indexBlock);
        System.out.println(name + " data blocks: " + fileBlocks);
    }

    // 🔹 Display Disk + Count
    static void displayDisk() {
        int free = 0, used = 0;

        System.out.print("Disk: ");
        for (int i = 0; i < totalBlocks; i++) {
            if (disk[i] == null) {
                System.out.print("- ");
                free++;
            } else {
                System.out.print(disk[i] + " ");
                used++;
            }
        }

        System.out.println();
        System.out.println("Total Blocks: " + totalBlocks);
        System.out.println("Used Blocks: " + used);
        System.out.println("Free Blocks: " + free);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        while (true) {
            System.out.println("\n1. Contiguous Allocation");
            System.out.println("2. Linked Allocation");
            System.out.println("3. Indexed Allocation");
            System.out.println("4. Display Disk");
            System.out.println("5. Exit");

            System.out.print("Enter choice: ");
            int choice = sc.nextInt();

            if (choice == 5) break;

            if (choice == 4) {
                displayDisk();
                continue;
            }

            System.out.print("Enter file name: ");
            String name = sc.next();

            System.out.print("Enter file size: ");
            int size = sc.nextInt();

            switch (choice) {
                case 1:
                    allocateContiguous(name, size);
                    break;
                case 2:
                    allocateLinked(name, size);
                    break;
                case 3:
                    allocateIndexed(name, size);
                    break;
                default:
                    System.out.println("Invalid choice");
            }
        }
    }
}